subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputPropertiesIntegrationTest.groovy

```diff
@@ -689,4 +690,101 @@ task someTask(type: SomeTask) {
+                int nestedInputValueCount = 0
+                private NestedBean bean = new NestedBean()
+                
+                @OutputFile
+                File getOutputFile() {
+                    count("outputFile", ++outputFileCount)
+                    return project.file('build/foo.bar')
+                }        
+                
+                @InputFile
+                File getInputFile() {
+                    count("inputFile", ++inputFileCount)
+                    return project.file('input.txt')
+                }
+                
+                @Input
+                String getInput() {
+                    count("inputValue", ++inputValueCount)
+                    return "Input"
+                }
+                
+                @Nested
+                Object getBean() {
+                    count("nestedInput", ++nestedInputCount)
+                    return bean
+                }
+                
+                @TaskAction
+                void doStuff() {
+                    outputFile.text = inputFile.text
+                }
+                
+                void count(String name, int currentValue) {
+                    println "Evaluating \${name} \${currentValue}"                
+                }
+                                
+                class NestedBean {
+                    @Input getFirst() {
+                        count("nestedInputValue", ++nestedInputValueCount)
+                        return "first"
+                    }
+                    
+                    @Input getSecond() {
+                        return "second"
+                    }
+                }
+            }
+            
+            task myTask(type: CustomTask)
+            
+            task assertInputCounts {
+                dependsOn myTask
+                doLast {
+                    ['outputFileCount', 'inputFileCount', 'inputValueCount', 'nestedInputCount', 'nestedInputValueCount'].each { name ->
+                        assert myTask."\$name" == project.property(name) as Integer                    
+                    }
+                }
+            }
+        """
+        def inputFile = file('input.txt')
+        inputFile.text = "input"
+        def expectedCounts = [inputFile: 3, outputFile: 3, nestedInput: 3, inputValue: 1, nestedInputValue: 1]
+        def expectedUpToDateCounts = [inputFile: 2, outputFile: 2, nestedInput: 3, inputValue: 1, nestedInputValue: 1]
+        def arguments = ["assertInputCounts"] + expectedCounts.collect { name, count -> "-P${name}Count=${count}"}
+        def upToDateArguments = ["assertInputCounts"] + expectedUpToDateCounts.collect { name, count -> "-P${name}Count=${count}"}
+
+        expect:
+        succeeds(*arguments)
+        executedAndNotSkipped(':myTask')
+
+        when:
+        inputFile.text = "changed"
+
+        then:
+        withBuildCache().succeeds(*arguments)
+        executedAndNotSkipped(':myTask')
```

Comment:
Why is it useful to run this with the cache enabled?

Reply:
We could be evaluating the properties more often when the cache is enabled/when loading from the cache (as we did before).